### PR TITLE
[C2B2/TimeLock Serialization] Phase 3b1: Vigor of the Breaking Dawn (Client Side Part 1)

### PR DESCRIPTION
**Pull request by jeremyk-91**
## General
**Before this PR**:
A client-side autobatcher that allows for users to make unlock calls for multiple namespaces, on the client side, does not exist. 

**After this PR**:

==COMMIT_MSG==
A client-side autobatcher that allows for users to make unlock calls for multiple namespaces, on the client side, exists. 
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**: No

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: No

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**: No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**: Yes

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**: Relies on timelock verison having #6170... but given the only user is AtlasDB Proxy, I believe it's less disruptive to just set the product dependency there.

**Does this PR need a schema migration?** No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**: No

**What was existing testing like? What have you done to improve it?**: Added tests for the batcher

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**: No

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**: No

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**: Not so much... other than seeing the endpoints called change.

**Has the safety of all log arguments been decided correctly?**: No new logging args.

**Will this change significantly affect our spending on metrics or logs?**: No

**How would I tell that this PR does not work in production? (monitors, etc.)**: Obvious 5xx-style errors.

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**: Rollback

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**: No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**: No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**: Eh. If we need lock tokens for something else, we'd change it accordingly.

## Development Process
**Where should we start reviewing?**: MultiClientTimeLockUnlocker is the heart of this PR.

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**: It's not.

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        public void applyResponse(ConjureUnlockResponseV2 relevantResponse) {
+            Set<LockToken> unlockedTokens = relevantResponse.get().stream()
+                    .map(conjureToken -> LockToken.of(conjureToken.get()))
+                    .collect(Collectors.toCollection(HashSet::new));
+            for (BatchElement<NamespacedLockSet, Set<LockToken>> batchElement : requests) {
+                Set<LockToken> plausibleUnlocks = ImmutableSet.copyOf(
+                        Sets.intersection(batchElement.argument().lockSet(), unlockedTokens));
+                batchElement.result().set(plausibleUnlocks);
+                unlockedTokens.removeAll(plausibleUnlocks);
+            }
```
Comment (sudiksha27) [inline lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:95]: is it never possible to have multiple requests to unlock same lock?

Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:95]: nvm just the saw the test

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.stream.Collectors;
+import org.immutables.value.Value;
+
+public class MultiClientTimeLockUnlocker implements AutoCloseable {
+    private final DisruptorAutobatcher<NamespacedLockSet, Set<LockToken>> batcher;
+
+    public MultiClientTimeLockUnlocker(InternalMultiClientConjureTimelockService delegate, OptionalInt bufferSize) {
+        this.batcher = Autobatchers.independent(new UnlockConsumer(delegate))
+                .bufferSize(bufferSize)
+                .batchFunctionTimeout(Duration.ofSeconds(30))
```
Comment (sudiksha27) [inline lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:50]: I think the unlock requests can be chonkier than other endpoints we have batched across namespaces. If the bufferSize chosen after taking into request/response size into account?

File: lock-api/src/test/java/com/palantir/lock/client/MultiClientTimeLockUnlockerTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    private static final ConjureLockTokenV2 CONJURE_TOKEN_3 = ConjureLockTokenV2.of(TOKEN_3.getRequestId());
+
+    private final InternalMultiClientConjureTimelockService conjureTimelockService =
+            mock(InternalMultiClientConjureTimelockService.class);
+    private final MultiClientTimeLockUnlocker unlocker =
+            new MultiClientTimeLockUnlocker(conjureTimelockService, OptionalInt.of(1024));
+
+    @Test
+    public void canUnlockOneUserRequest() {
+        when(conjureTimelockService.unlock(
```
Comment (sudiksha27) [inline lock-api/src/test/java/com/palantir/lock/client/MultiClientTimeLockUnlockerTest.java:60]: can we add a test for exception handling if the server throws? Also, is partial response possible?

**Review by sudiksha27 — COMMENTED** [review]
Almost there

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            for (Map.Entry<Namespace, SingleClientBatchManager> batchManagerEntry : batchManagers.entrySet()) {
+                ConjureUnlockResponseV2 relevantResponse = responses.get(batchManagerEntry.getKey());
+                batchManagerEntry.getValue().applyResponse(relevantResponse);
+            }
+        }
+    }
+
+    @Value.Immutable
+    interface NamespacedLockSet {
+        @Value.Parameter
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:130]: Yeah, that's fine - changed.

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                NamespacedLockSet namespacedLockSet = batchElement.argument();
+                batchManagers
+                        .computeIfAbsent(namespacedLockSet.namespace(), unused -> new SingleClientBatchManager())
+                        .addBatchElement(batchElement);
+            }
+
+            Map<Namespace, ConjureUnlockResponseV2> responses = timelockService.unlock(KeyedStream.stream(batchManagers)
+                    .map(SingleClientBatchManager::getCombinedRequest)
+                    .collectToMap());
+
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:120]: I think we can wait until the internal proxy PR though, right? Merging this (without modifying the batchers) should still use the legacy endpoint, unless I've got something wrong

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                batchElement.result().set(plausibleUnlocks);
+                unlockedTokens.removeAll(plausibleUnlocks);
+            }
+        }
+    }
+
+    @VisibleForTesting
+    static class UnlockConsumer implements Consumer<List<BatchElement<NamespacedLockSet, Set<LockToken>>>> {
+        private final InternalMultiClientConjureTimelockService timelockService;
+
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:102]: Not really, other than I found this easier to read/understand. I'm happy to change it to match the others if you think that would be appropriate.

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java (hunk 2 above)
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:50]: Yes, that's a valid flag. Though that's probably something on the AtlasDB Proxy side to decide: a lock token is a UUID and so 36 bytes in string form; let's round that up to 40 to account for json overheads. So if we have a 50 MB limit on requests, we can get away with 1.25MM lock tokens. While it's hard to bound the size of an individual request, 16,384 would mean a given unlock request can have about 76 tokens; we can choose a lower constraint for this batcher and I'd think even say 1,024 for the ring buffer size would already give us some improvements.

File: lock-api/src/test/java/com/palantir/lock/client/MultiClientTimeLockUnlockerTest.java (hunk 3 above)
Comment (jeremyk-91) [reply lock-api/src/test/java/com/palantir/lock/client/MultiClientTimeLockUnlockerTest.java:60]: Partial response shouldn't be possible based on the server implementation (see #6170). I'll add a test for exception handling.

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:120]: yeah sure that makes sense, as long as we have not wired it in this is fine

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:102]: this is pretty minor so this is fine. I leave it up to you.

File: lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java (hunk 2 above)
Comment (sudiksha27) [reply lock-api/src/main/java/com/palantir/lock/client/MultiClientTimeLockUnlocker.java:50]: Thanks! That makes sense. As long as we do not fall back on default which I _think_ is 4096 in internal proxy (which also seems like a reasonable estimate)

**Review by sudiksha27 — APPROVED** [review]
lgtm!

**Pull request merged**